### PR TITLE
Stop using short data store IDs in detached container

### PR DESCRIPTION
**Pull request by agarwal-navin**
## Bug
Currently, data stores created in detached container always get a short ID via id compressor. We have a bug in one of our customer's app where a short data store ID created is `[` but in a downloaded snapshot, it is converted to its ASCII equivalent `%5B` in certain conditions. So, when an op comes for this data store with id `[`, containers loaded with this snapshot cannot find the data store.

## Root-cause
We haven't yet identified what causes the id to change from `[` -> `%5B`. However, we know that it is caused by short IDs generated in detached container.

## Fix
To mitigate the issue for customers, this PR disables the generation of short IDs for data stores and DDSes. Added tests that validate that we don't generate short IDs for data stores and DDSes in detached or attached state.

### CODE REVIEW

File: packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts
Comment (vladsud) [inline packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts:949]: Why it's unknown, not boolean type? All the cases seem to be boolean.

File: packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts
Comment (vladsud) [inline packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts:999]: Maybe be roll it into previous assert? I.e. 

assertInvert(typeof idTest === "number" && idTest >= 0, "short IDs should be issued");

File: packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts
Comment (vladsud) [inline packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts:1068]: Should we change this test and create 100 of detached data stores?

**Review by vladsud — APPROVED** [review]

File: packages/runtime/datastore/src/dataStoreRuntime.ts
Comment (rajatch-ff) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:453]: Should we put these changes behind a feature flag, for easy toggles in DF/PROD?

**Comment by rajatch-ff** [comment]
Any side-effects of turning off short-ids that we should know about or consider before merging?

File: packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts:1068]: We could but I don't see a point. Are you suggesting we the load a container and basically validate that it can load all these data stores, process ops, etc?

File: packages/runtime/datastore/src/dataStoreRuntime.ts
Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:453]: I am not sure I see the value in doing that because we can't enable it unless we fix the bug which will require making code changes anyway. We can enable it via that change and add feature flags if needed.

File: packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts]: Changed it to boolean

File: packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts]: Done

File: packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts:1067]: Yeah you need 13 to create a datastore with id "["

**Comment by agarwal-navin** [comment]
> Any side-effects of turning off short-ids that we should know about or consider before merging?

AFAIK, there shouldn't be. I will defer to @vladsud to double check.

File: packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts:1067]: Sure, but there is no repro of the issue even with that. So, I don't see a point in doing it yet. Once we have a repro, we can add test for this.

Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts:1067]: Working on it

File: packages/runtime/datastore/src/dataStoreRuntime.ts
Comment (rajatch-ff) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:453]: I hear you there. Feature flag might come handy, if the issue happens to be on ODSP side, and we can just toggle our change off without making a code change. Since the true-root-cause is still being figured out, there may be value in keeping this change behind a switch.

**Comment by vladsud** [comment]
We are not attempting to recover existing documents. Change impacts new documents only (it will make them work again).

File: packages/runtime/datastore/src/dataStoreRuntime.ts
Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:453]: I can add it behing a feature. But I doubt we will use it (Loop would probably not let us it being so risky). Also, we heard from ODSP that it's not their issue most likely.
Let me work on adding a feature, it doesn't hurt :).

Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts]: Added feature flags

**Review by rajatch5 — APPROVED** [review]

File: packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts:1067]: https://github.com/microsoft/FluidFramework/pull/21625/files

**Pull request merged**
